Multithreaded implementation

netj opened the pull request:
Lays down the foundation for #3.  This version will be used for fixing crashes on latest Mac (10.11) caused by nonblocking I/O implementation (HazyResearch/deepdive#522).

@alldefector can you take a look at the code, perhaps just the [mkmimo_multithreaded.c](https://github.com/netj/mkmimo/pull/8/files#diff-a1e439a448f4775c4138223c606c8ce0) if the rest seems too noisy?

sashaperigo committed:
Mulithreaded progress thus far. Working on condition variables

netj committed:
make format

netj committed:
Adds multithreaded-friendly interfaces to queue

that uses condition variable with mutex

netj committed:
Implementing multithreaded version

netj committed:
Finishes multi-threaded version

- not forgetting to unlock, initialize mutex and cond var
- completing trailing data copy function
- introducing flush thread to ensure output threads are waken up
  afterwards

netj committed:
Makes perrorf reentrant

netj committed:
Cleans up #includes in headers

clang-format 3.8 compatible order

netj committed:
Fixes warning on Linux with strerror_r

and turns all warnings to errors

netj committed:
Switches flags to bool type

netj committed:
Relies on compiler generated dependencies

netj committed:
Lets make expand wildcards

netj committed:
Factors out a buffer utility

for moving trailing data after the last record to another buffer

netj committed:
Adds travis env matrix to test both implementations

netj committed:
Adds DEBUG=1 to travis test matrix

netj committed:
Sets GCC version to 5.3 on Travis

netj committed:
Improves regex to capture num bytes from dd output

and mutes DEBUG output when measuring throughput

netj committed:
Corrects gcc-5 package/command name

netj committed:
Allows DEBUG build to fail performance test

netj committed:
Installs gcc-5 only when necessary

netj committed:
Allows DEBUG build to fail another sensible output test

netj committed:
Fixes test bug skipping even normal ones as DEBUG build

netj committed:
Adds MULTIBUFFERING parameter

for determining multiple-buffering factor

netj committed:
Adds a bunch of info to README

usage, parameters, etc.

netj committed:
Merge branch 'master' of https://github.com/netj/mkmimo into multithreaded

netj committed:
Updates README about bash implementation

netj committed:
Revises some comments

alldefector commented on mkmimo_multithreaded.c:
```diff
@@ -1,15 +1,310 @@
+
+      if (num_bytes_read < 0) {
+        // Close input upon errors
+        perrorf("read %s", input->name);
```
Message is not clear that there is an IO error. Maybe also print out `num_bytes_read` if the code is meaningful.

netj replied:
`perrorf` is a macro that appends the actual error string that corresponds to the last error number, so I think it'll look clear.  Actually, I realize two issues:
1. I think it currently relies on errno global, so I probably need to make it reentrant/threadsafe.
2. Actually these errors are not escalating into a non-zero exit status, then will surely get silenced, which will be really awful.

Thanks for point me to this direction :)

alldefector commented:
Looks great! How does its performance compare with nonblocking?

netj commented:
@alldefector Thanks for taking a look!  Haven't done any real measurement, but it easily passes throughput tests on Mac, which wasn't the case for nonblocking impl.  So I'll handle the error exit status and also measure things a bit.  I'm hopeful this will saturate system's data bandwidth, e.g., ~100GB/s for raiders as I measured before (#3).

Btw while coding this, I was running nonblocking impl. on an iMac with 10.10, keeping it busy at 300-400MB/s, but it never rebooted once.  Hilariously, my laptop rebooted while writing this comment as I was waiting for the first deepdive make test to finish on my fresh MacBook Pro.. and I'm now typing it again from scratch.  We need to finish this asap and fix deepdive.. :)

netj committed:
Escalates read(2) error to non-zero exit status

However, I failed to create a test that generates error for
read(2)/write(2) syscalls.

See: https://gist.github.com/netj/8f7e2e73f3913d12703e

netj commented:
`errno` turns out to be already thread-safe, and I made sure the code exits with non-zero status upon I/O errors but failed to create a precise test.

When I tried a very rough throughput test with something like below, the new multithreaded showed 2-5x improvement in throughput, each channel going 40MB/s (nonblocking) -> 100-250MB/s (multithreaded) for 30 I/O streams.  But there are lots of knobs to turn here, so not clear where these two shine.  @sashaperigo and I will start exploring this next quarter.  Anyway, the conclusion is the new implementation often delivers higher throughput, doesn't crash Mac. :)

``` bash
xMKMIMO_IMPL=nonblocking \
MULTIBUFFERING=2 \
PATH=$PATH:$PWD/test/util \
mkmimo_throughput timeout=1m \
num_producer_fast=50 \
num_producer_slow=10 \
num_consumer_fast=50 \
num_consumer_slow=10 \
producer_fast_throughput= \
producer_slow_throughput=2k \
consumer_fast_throughput= \
consumer_slow_throughput=1k \
#
```

One last thing that concerns me is a potential deadlock issue.  I observed tests hanging either at the beginning or at the end, and I think I'll have to do some stress tests to make sure it's completely reliable.  If you can, please take a look at some obvious parts that play with mutex and embarrass me. :)

netj committed:
Always CHECK_ERRNO in Queue and use errorchecking mutex

netj committed:
Increases the size of tests

netj committed:
make format

netj committed:
Fixes hanging bug by keeping flushing thread indefinitely

circulating empty buffers so all output threads finish voluntarily eventually

netj committed:
Reverts to normal mutex (without error checking)

This reverts part of commit 5ed2d6b561df71998754c2808237b3db84231c5a.

Conflicts:
	queue.c

alldefector commented:
Seems good to go!
